src/Microsoft.AspNetCore.Blazor.Browser/Hosting/BrowserHost.cs

```diff
@@ -0,0 +1,101 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Blazor.Hosting
+{
+    internal class BrowserHost : IBrowserHost
+    {
+        private readonly IJSRuntime _runtime;
+
+        private IServiceScope _scope;
+        private BrowserRenderer _renderer;
+
+        public BrowserHost(IServiceProvider services, IJSRuntime runtime)
+        {
+            Services = services ?? throw new ArgumentNullException(nameof(services));
+            _runtime = runtime ?? throw new ArgumentNullException(nameof(runtime));
+        }
+
+        public IServiceProvider Services { get; }
+
+        public Task StartAsync(CancellationToken cancellationToken = default)
+        {
+            // We need to do this as early as possible, it eliminates a bunch of problems. Note that what we do
+            // is a bit fragile. If you see things breaking because JSRuntime.Current isn't set, then it's likely
+            // that something on the startup path went wrong.
+            //
+            // We want to the JSRuntime created here to be the 'ambient' runtime when JS calls back into .NET. When
+            // this happens in the browser it will be a direct call from Mono. We effectively needs to set the
+            // JSRuntime in the 'root' execution context which implies that we want to do as part of a direct
+            // call from Program.Main, and before any 'awaits'.
+            JSRuntime.SetCurrentJSRuntime(_runtime);
+
+            var scopeFactory = Services.GetRequiredService<IServiceScopeFactory>();
+            _scope = scopeFactory.CreateScope();
+
+            try
+            {
+                var startup = _scope.ServiceProvider.GetService<IBlazorStartup>();
+                if (startup == null)
+                {
+                    var message =
+                        $"Could not find a registered Blazor Startup class. " +
+                        $"Using {nameof(IBrowserHost)} requires a call to {nameof(IBrowserHostBuilder)}.UseBlazorStartup.";
+                    throw new InvalidOperationException(message);
+                }
+
+                // Note that we differ from the WebHost startup path here by using a 'scope' for the app builder
+                // as well as the Configure method.
+                var builder = new BrowserBlazorApplicationBuilder(_scope.ServiceProvider);
+                startup.Configure(builder, _scope.ServiceProvider);
+
+                _renderer = builder.CreateRenderer();
+            }
+            catch
+            {
+                _scope.Dispose();
+                _scope = null;
+
+                if (_renderer != null)
+                {
+                    _renderer.Dispose();
+                    _renderer = null;
+                }
+
+                throw;
+            }
+
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken = default)
```

Comment:
I'm guessing this never actually happens in the browser case. Or is there any way the developer can trigger it?

Reply:
No there's no way to trigger it right now. 

The purpose of this (and reason to leave it in) would be if we want to expose the ability for services to register that want to know the application is shutting down. This would be the place that we trigger that logic.